src/Compilers/Core/CodeAnalysisTest/Collections/TopologicalSortTests.cs

```diff
@@ -0,0 +1,204 @@
+        [InlineData(1075862430)]
+        [InlineData(428872484)]
+        [InlineData(489337268)]
+        [InlineData(1976108951)]
+        [InlineData(428397397)]
+        [InlineData(1921108202)]
+        [InlineData(926330127)]
+        [InlineData(364136202)]
+        [InlineData(1893034696)]
+        public void TestRandom(int seed)
+        {
+            int numberOfNodes = 100;
+            Random random = new Random(seed);
+
+            // First, we produce a list of integers representing a possible (reversed)
+            // topological sort of the graph we will construct
+            var possibleSort = Enumerable.Range(0, numberOfNodes).ToArray();
+            shuffle(possibleSort);
+
+            // Then we produce a set of edges that is consistent with that possible topological sort
+            int[][] successors = new int[numberOfNodes][];
+            for (int i = numberOfNodes-1; i >= 0; i--)
+            {
+                successors[possibleSort[i]] = randomSubset((int)Math.Sqrt(i), i);
+            }
+
+            // Perform a topological sort and check it.
+            var sorted = TopologicalSort.IterativeSort<int>(Enumerable.Range(0, numberOfNodes).ToArray(), x => successors[x]);
+            Assert.Equal(numberOfNodes, sorted.Length);
+            AssertTopologicallySorted(sorted, successors, $"TestRandom(seed: {seed})");
+
+            // Now we modify the graph to add an edge from the last node to the first node, which
+            // probably induces a cycle.  Since the graph is random, it is possible that this does
+            // not induce a cycle. However, by the construction of the graph it is almost certain
+            // that a cycle is induced. Nevertheless, to avoid flakiness in the tests, we do not
+            // test with actual random graphs, but with graphs based on pseudo-random sequences using
+            // random seeds hardcoded into the tests. That way we are testing on the same graphs each
+            // time.
+            successors[possibleSort[0]] = successors[possibleSort[0]].Concat(new int[] { possibleSort[numberOfNodes - 1] }).ToArray();
+
+            Assert.Throws<ArgumentException>(() =>
+            {
+                TopologicalSort.IterativeSort<int>(Enumerable.Range(0, numberOfNodes).ToArray(), x => successors[x]);
+            });
+
+            // where
+            void shuffle(int[] data)
+            {
+                int length = data.Length;
+                for (int t = 0; t < length - 1; t++)
+                {
+                    int r = random.Next(t, length);
+                    if (t != r)
+                    {
+                        var tmp = data[t];
+                        data[t] = data[r];
+                        data[r] = tmp;
+                    }
+                }
+            }
+
+            int[] randomSubset(int count, int limit)
+            {
+                // We don't worry about duplicate values. That's all part of the test,
```

Comment:
💭 I'm not sure this is correct. If you intend to support duplicate values in the result of `successors`, you should add a dedicated test to that effect.

Reply:
The predecessor count is a count of incoming edges, not preceding nodes. I'll add a test.

---
In reply to: [171728301](https://github.com/dotnet/roslyn/pull/25169#discussion_r171728301) [](ancestors = 171728301)